docs/Memory.md

```diff
@@ -1,34 +1,379 @@
 # Memory Management - a closer look
 
-### Related pages
+Carp uses a *linear type system* to manage the memory associated with different
+values throughout a program. Carp's memory management system is designed and
+implemented with the following goals in mind: 
+
+* Predictable: The memory management system's behavior should be easy to reason
+  about.
+* Efficient: The memory management system should not have significant impacts
+  on performance.
+* Safe: The memory management system should prevent errors related to memory
+  management, such as "use after free" and "double free"
+
+This document introduces the basic concepts behind the kind of linear type
+system Carp uses. In addition, it takes a deeper look at how the system is
+currently implemented.
+
+## Linear Types and Memory Management
+
+Carp's linear type system tracks the *ownership* of the memory associated with a
+given value as part of its type signature. A *linear type* is a traditional type
+with additional information called a *lifetime* that allows the type system to
+track a value's association with a given memory location.
+
+The memory management system *only* manages linear types; not all types are
+linear. Some of Carp's builtin types are linear by default:
+
+- The String type is linear and managed by the memory system.
+- The Pattern type is linear and managed by the memory system.
+- The Array type is linear and managed by the memory system.
+- The Box type is linear and managed by the memory system.
+- Function types are linear and managed by the memory system.
+
+All other builtin types are *not* linear, and thus aren't managed by the memory
+system.
+
+A few conditions determine whether or not a user defined type is linear:
+
+- **Implementation of the `blit` interface**: this interface explicitly marks a
+  type as *non-linear*. Any type that implements it is ignored by the memory
+  management system and is assumed to pose no risks in relation to memory
+  allocation and deallocation.
+- **Implementation of the `delete` interface**: this interface explicitly marks
+  a type as *linear*. Any type that implements it is managed by the memory
+  management system. Carp will call the implementation of this interface whenever
+  the memory management system decides it's safe to deallocate the memory
+  associated with a value of the type that implements this interface. 
+
```

Comment:
Should there be a mention that product and sumtype will auto-implement `delete` somewhere?

Reply:
Great addition! added, thanks! I think we should eventually mention `drop` somewhere in this doc as well, but I'll leave that for a future PR.